base/path.jl

```diff
@@ -163,17 +163,39 @@ See also: [`basename`](@ref)
 """
     basename(path::AbstractString) -> AbstractString
 
-Get the file name part of a path.
+Get the file/dir name part of a path.
+
+Generally, trailing '/' will be ignored. There're some other exceptions:
+
+* ""  ===> "."
+* "~" ==> basename(ENV["HOME"])
+* "/" ==> "/"
+* "//" => "/"
 
 # Examples
 ```jldoctest
 julia> basename("/home/myuser/example.jl")
 "example.jl"
+julia> basename("/opt/")
+"opt"
 ```
 
 See also: [`dirname`](@ref)
+
+Reference: [POSIX basename](http://pubs.opengroup.org/onlinepubs/9699919799/functions/basename.html)
 """
-basename(path::AbstractString) = splitdir(path)[2]
+function basename(path::AbstractString)
+    if path == ""
+        return "."
+    elseif path == "/"
+        return "/"
+    elseif path == "~"
+        return basename(ENV["HOME"])
```

Comment:
POSIX `basename()` doesn't do ~ expansion and I don't think `ENV["HOME"]` will work on Windows.  I don't think ~ expansion should be done here.

Reply:
Not to mention ~ is a valid filename...

Reply:
Thanks for the review, I'm really not very familiar with these rules.

I add this because I checked the `basename ~` in Linux and macOS, both returns `jc` in my case. And `basename ~/~` returns `~`

Reply:
When running basename from the command line it is the shell that is doing the ~ expansion, not the basename utility.  Try `basename '~'`.